### ISSUE TITLE
Setups are not working as expected after upgrade

### ISSUE THREAD

**Issue by Naxemar** [devlooped/moq#1152]
Hi everyone, from time to time we are performing a regular nuget update, and currently part of our previously working tests are not working any more. We have tried several times with different versions of a package and also we have waited for a fix but it never came. At the beginning we were not able to find a cause and to reproduce it. We have also seen similar issues closed because of inability to reproduce. Eventually though we have manage to reproduce it. Bug was introduced with Moq 4.14.0 and still exists. Last correctly working version was 4.13.1. 

Steps to reproduce:

```
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Linq;
using Moq;

namespace MoqIssueTestProject
{
    public interface ITest1Interface
    {

    }

    public interface ITest2Interface : ITest1Interface
    {

    }

    public interface ICreatorInterface
    {
        T Create<T>(object[] arguments);
    }

    public class ClassToTest
    {
        private readonly ICreatorInterface creator;
        
        public ClassToTest(ICreatorInterface creator)
        {
            this.creator = creator;
        }

        public ITest1Interface[] ConstructedInstances { get; private set; }

        public void MethodToTest()
        {
            this.ConstructedInstances = Enumerable.Range(0,5).Select(x => this.creator.Create<ITest2Interface>(null)).Cast<ITest1Interface>().ToArray();
        }
    }

    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void TestMethod1()
        {
            // Arrange
            var mockCreator = new Mock<ICreatorInterface>();
            mockCreator.Setup(x => x.Create<ITest2Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest2Interface>().Object);
            mockCreator.Setup(x => x.Create<ITest1Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest1Interface>().Object); // this breaks things
            var testCandidate = new ClassToTest(mockCreator.Object);

            // Act
            testCandidate.MethodToTest();

            // Verify
            CollectionAssert.AllItemsAreInstancesOfType(testCandidate.ConstructedInstances, typeof(ITest2Interface));

        }
    }
}

```

**Comment by stakx** [comment]
Hi @Naxemar, thanks for the complete report.

TL;DR: switch the order of your two setups so they're ordered "less specific" to "more specific".

Your setups are logically in the wrong order, they should be ordered from "least specific setup" to "most specific setup"... otherwise the less specific setup will shadow/override the more specific setup; Moq always chooses the latest matching setup.

By specificity, I am in your case referring to the fact that a test such as `x is ITest2Interface` yields true only when `x` is a `ITest2Interface`, but not when `x` is a `ITest1Interface`. Contrast this with `x is ITest1Interfae`, which yields true in both cases.

In other words, your setup for `ITest1Interface` also covers those calls targeted by the setup for `ITest2Interface`, the latter is therefore the more specific one and should go last.

Yes, there was a change in behavior in past Moq versions regarding this &ndash; I'm suspecting because of #1004 that went into 4.14.0, in conjunction with #904 that went into an earlier release &ndash; but since the new (intended!) behavior is more consistent with how Moq works generally, this was a correction rather than a regression, so we won't be changing things back to the previous behavior.

**Comment by Naxemar** [comment]
Thank you @stakx for quick response, I understand. In our case it was not that obvious as some preparations are done in the initialize / base initialize / helpers and some preps are rather case specific and now it means that for such scenarios we need to duplicate the test preparations to keep Moq expected prep order. Sounds strange but if it must stay like that then we can handle that as well. 

Edit:
Maybe one thing to consider would be to support overriding a setup at the correct place, when it happens later for exact preparation, example:

```var mockCreator = new Mock<ICreatorInterface>();
mockCreator.Setup(x => x.Create<ITest1Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest1Interface>().Object);
mockCreator.Setup(x => x.Create<ITest2Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest2Interface>().Object);
// expected order, right

(...)

mockCreator.Setup(x => x.Create<ITest1Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest1Interface>().Setup(...).Returns(...).Object); 
// this will now brak the ITest2Interface preparation

**Comment by Artemigos** [comment]
My knowledge is a little surface level here, but something feels wrong.

I don't like the assumption that `Create<ITest1Interface>` is a more general version of `Create<ITest2Interface>`. It's true for the interface types, but not for the methods that use those types as type parameters.

The important detail here is that generic type parameters are an equivalent of taking `Type` as an input parameter. If we transcribe the example we get:

```csharp
object Create(Type returnType, object[] arguments);
```

In that case it's clear that doing a setup with `typeof(ITest1Interface)` is different from a setup for `typeof(ITest2Interface)` and the first one cannot be used to handle invocations with `typeof(ITest2Interface)` as an argument. I think this rule applies for generic type parameters as well.

**Comment by Naxemar** [comment]
Maybe one thing to consider would be to support overriding a setup at the correct place, when it happens later for exact preparation example:

```var mockCreator = new Mock<ICreatorInterface>();
mockCreator.Setup(x => x.Create<ITest1Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest1Interface>().Object);
mockCreator.Setup(x => x.Create<ITest2Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest2Interface>().Object);
// expected order, right

(...)

mockCreator.Setup(x => x.Create<ITest1Interface>(It.IsAny<object[]>())).Returns(new Mock<ITest1Interface>().Setup(...).Returns(...).Object); 
// this will now brak the ITest2Interface preparation

**Comment by stakx** [comment]
@Artemigos, the `Type` analogy isn't useful here, because we're talking only about static (i.e. compile-time) types. `Type` is an entirely different beast, it carries type information as data at runtime. Not really applicable here at all, but maybe I misunderstand what you're trying to say.

**Comment by Artemigos** [comment]
I'm trying to compare this on conceptual level, maybe it was a little misleading.

The point I'm trying to make is this - generic methods take the types as input and run logic based on it, just like normal methods run logic based on values that were passed in. I can easily think of cases where `Method<BaseType>(...)` should give a different result from `Method<DerivedType>(...)`:

- `GetTypeName<T>()`
- `IsInLayer<T>()` - which checks in which layer a type resides, based on an attribute on the class
- `CanSerialize<T>()` - static method defines if type can be serialized, base class allows, derive class doesn't
- the strongest example `iocContainer.Resolve<T>` - calling `Resolve<IInterface>()` has a **very** different meaning than `Resolve<ClassThatImplementsInterface>()` - in the first case it looks for a specific type implementing it and instantiates it, the second one goes straight to instantiation

So my conclusion here is that the generic type parameter is an input like any other and making assumptions that the setup should match for different input types seems dangerous.

Also, I'm starting to feel that I'm misunderstanding something. What is going to be the result of the code below?

```csharp
public interface IFactory
{
    public T Create<T>() { ... }
}

public interface IInterface { ... }
public class Derived1 : IInterface { ... }
public class Derived2 : IInterface { ... }

var mock = new Mock<IFactory>().Setup(x => x.Create<IInterface>()).Returns(new Derived1());
Derived2 d2 = mock.Object.Create<Derived2>();
```

**Comment by stakx** [comment]
@Artemigos your last code example should throw. (I suggest you run it and see for yourself.) This is only possible because Moq's setup API isn't fully type-safe.

I think you're mainly missing the fact that generic type arguments aren't inputs (in the sense of a constant value) like regular arguments. They are types and as such polymorphism needs to be considered. When you have a setup like this one:

```csharp
mock.Setup(m => m.Method<Foo>())
```

the `Foo` doesn't mean that you only setup `Method` for type argument `Foo`; you set up `Method` for any type argument that is assignment compatible to `Foo`. Generic type arguments in setup expressions restrict a type parameter kind of similar to a `where T : Foo` constraint. Polymorphism / subtype relationships must be observed.

```csharp
public interface IPerson
{
    void Eat<T>(T thing);
}

class Apple { }
class GrannySmithApple : Apple { }

var fred = new Mock<IPerson>();
fred.Setup(f => f.Eat(It.IsAny<Apple>())).Callback(() => WriteLine("Fred is making happy munch munch noises."));
fred.Setup(f => f.Eat(It.IsAny<GrannySmithApple>())).Callback(() => WriteLine("Fred just lost a tooth.");

fred.Object.Eat(new GrannySmithApple());
```

Moq favours more recent setups over earlier ones, so it first checks whether you're calling `Eat<T>(T thing)` where `T` is a subtype of `GrannySmithApple`; you are, so it chooses that setup and displays the tooth message.

If you swap the setup order, Moq will again begin with the last setup and check whether you're calling `Eat<T>(T thing)` where `T` is a subtype of `Apple`;  you are (because a `GrannySmithApple` is an `Apple`) so it chooses that setup and you see the munch munch message.

The above Moq setups corresponds roughly to this manual mock:

```csharp
class FredMock : IPerson
{
    public void Eat<T>(T thing)
    {
        if (thing is GrannySmithApple)
        {
            WriteLine("...tooth.");
        }
        else if (thing is Apple)
        {
            WriteLine("...munch munch...");
        }
    }
}
```

Moq used to treat generic type arguments in setup expressions as precise constants (I'm simplifying a bit here), but that didn't match the usual C# language semantics very well. If we hadn't changed that, and say you only had the setup for `Eat(It.IsAny<Apple>())` but made a call with a `GrannySmithApple`, then the setup wouldn't have triggered because `GrannySmithApple` doesn't precisely match `Apple`; not very intuitive. Therefore the changes towards polymorphic behavior in 4.13.0 and 4.14.0.

**Comment by Artemigos** [comment]
> your last code example should throw

This supports my argument I think.

Polymorphism applies to types, not methods. Also, for generic types, polymorphism doesn't apply without the `in` and `out` keywords.

```csharp
public interface IInt<T> {}
public class Cls<T> : IInt<T> {}

public class A {}
public class B : A {}
public class C : B {}

var a = new Cls<B>();
IInt<A> cast1 = a;
IInt<C> cast2 = a;
```

With this setup I can add `in` or `out` to the `T` on the interface. If I add `in` only casting `IInt<B>` to `IInt<C>` is allowed. If I add `out` only casting `IInt<B>` to `IInt<A>` is allowed. If I don't add anything both casts are **not** allowed by the compiler.

It is not possible to add `in` or `out` on type parameters that appear on methods, but we can try to create a delegate type, that would work for that method and allow casting. If we do that, then **for some methods** it's possible to create a delegate that will allow casting in various directions, but the covariance/contravariance has to be inferred or ignored - first of all this allows type safety breakage like shown above, but second of all it assumes that this is always what the developer intends - and this is the part that worries me.

As for how it breaks:

```csharp
public T Create<T>() { ... }

mock.Setup(x => x.Create<B>()).Returns(...);
```

The setup takes `Func<B>` and then when there's a call to `Create<C>()` it tries to cast the function **down** to `Func<C>` to call it. The C# compiler doesn't allow this cast when it's statically typed.

```csharp
Func<B> fb = Create<B>;
Func<C> fc = fb; // compiler error
```

(BTW I'm aware that through the magic of dynamic programming you probably don't cast anything, this is about what happens 'in theory')

**Comment by stakx** [comment]
@Artemigos:

> ```csharp
> public T Create<T>() { ... }
> ```

Just for completeness' sake, you probably meant:

```csharp
public interface IWillBeMocked<out T>
{
    T Create<T>();
}
```

And beyond that, it's your responsibility to ensure that your setup's  `.Returns(...)` will produce a value of the correct (compatible) type, because (like I said above) Moq 4's setup API isn't (and IIRC unfortunately cannot be made) fully type-safe.

**Comment by stakx** [comment]
@Naxemar:

> Maybe one thing to consider would be to support overriding a setup at the correct place, when it happens later for exact preparation

I'm not sure I understand your suggestion.

Is it the augmented chaining `x.Setup(...).Returns(...).Setup(...).Returns(...).Object` you're after? If so, we probably cannot do that without a breaking chance in the public API. What's probably worse is that it makes it less clear when you're configuring a setup, and when you're configuring the mock. Perhaps better not do allow this, the practical benefit is quite small anyway (i.e. you don't save a lot of characters compared to `x.Setup(...).Returns(...)` followed by another `x.Setup(...).Returns(...)`.

Besides that, I'd say it isn't terribly good practice to "neutralise" or cancel existing setups by re-layering the same setups on top but in a different order. It may be better to simply recreate a fresh mock, in order to keep your test code as free from unnecessary logic and state as possible.

I hope you don't mind me closing this issue; I don't think there is any immediate reason to change something about Moq here. If I have misunderstood your suggestion, please do post again!
